Un-inline the enums in the CST

Xanewok opened the pull request:
Part of https://github.com/NomicFoundation/slang/issues/638

As we discussed, inlining some nodes types causes us to lose information that is otherwise hard/inconvenient to reconstruct and this hopefully increases the usability of CST alone.

OmarTawfik commented on crates/solidity/testing/snapshots/cst_output/ContractDefinition/recovery_testbed/generated/0.5.0-failure.yml:
```diff
@@ -47,62 +47,78 @@ Tree:
+                          - Statements (Rule): # 157..174 "\t\tuint256 x = 0;\n"
+                              - Statement (Rule): # 157..174 "\t\tuint256 x = 0;\n"
+                                  - VariableDeclarationStatement (Rule): # 157..174 "\t\tuint256 x = 0;\n"
+                                      - VariableDeclarationType (Rule): # 157..166 "\t\tuint256"
+                                          - TypeName (Rule): # 157..166 "\t\tuint256"
+                                              - ElementaryType (Rule): # 157..166 "\t\tuint256"
+                                                  - UintKeyword (Token): "uint256" # 159..166
```
One idea we discussed before when we talked about this is flattening the tree structure of the tests a little bit to make it more readable/usable. So, for the below example:

- Instead of each node on a separate line, we can group rules that have only one child (that is also a rule) into the same line, as they will have the same exact range and preview comment.
- Removing `(Rule)` and `(Token)` suffixes, since we already make the dinstinction through the YAML value/LHS on the same line.

Just putting the idea out here. Definetely not blocking for this PR, as we probably should do it in a subsequent PR to make it easier to review.

```suggestion
                                      - VariableDeclarationType -> TypeName -> ElementaryType: # 157..166 "\t\tuint256"
                                          - UintKeyword: "uint256" # 159..166
```

OmarTawfik replied:
I'll add a TODO for myself.

OmarTawfik commented on crates/solidity/inputs/language/src/grammar.rs:
```diff
@@ -322,8 +322,7 @@ fn resolve_grammar_element(ident: &Identifier, ctx: &mut ResolveCtx) -> GrammarE
                 context: lex_ctx,
-                // Enums have a single reference per variant, so they should be inlined.
-                is_inline: matches!(elem.as_ref(), Item::Enum { .. }),
+                is_inline: false,
```
I would like to get @AntonyBlakey's eyes on this one. Some of these nodes look very useful, for example `ContractMember` and `ConstructorAttribute`, as it makes it easier to find/match on these elements.

However, some of them look extraneous indeed, and are only there for purpose of authoring the grammar/versioning. For example:

- `ElementaryType` and `YulLiteral` which will (almost) always have a unique parent that can be matched against. Maybe we can refactor the grammar a bit to make this more accurate?

- `TypedTupleMember` and `UntypedTupleMember` which only exist to make parsing/backtracking correct, but provide no additional meaning. Not sure how to avoid it.

What do you think?

I’m trying to avoid adding optional inlining to the DSL unless we absolutely need it. As based on all of our new design decisions, and AST structure, it will make it less obvious to go from grammar to CST/AST, and add another layer of complexity that users have to deal with ..

So, if we are happy with these changes, I’m fine with merging the PR as-is for now, and I can manually go over the enums to see if any of them can be better structured. For example, inlining something like `ElementaryType` variants into the types it references, since it is almost always used inside another Enum. It will probably be more nuanced than that though.

AntonyBlakey replied:
Could we make the non-interesting enum into something else in the grammar i.e. effectively add a non-outlined characteristic by using a different name for the concept. It sure sounds like it has a different purpose.

OmarTawfik replied:
f2f: we think the PR is good enough for now, and let's review the added kinds later to see if anything can be pruned/removed.

Xanewok commented:
Rebased and force-pushed; this is still just https://github.com/NomicFoundation/slang/pull/698/commits/c4c2f4afd623fc2c4f972e4003d71bfb98f85ec2 + re-generated files and adjusted test.
